File: Java/src/_0027_Remove_Element/Solution.java
package _0027_Remove_Element;

class Solution {
    // 时间复杂度：O(n)
    // 空间复杂度：O(1)
    public int removeElement(int[] nums, int val) {
        int k = 0;  // nums[0...k)中的元素均为删除val后的元素

        for (int i = 0; i < nums.length; i++)  // 两个索引，一个跟踪当前考察的元素，另一个考察删除val最新的索引
            if (nums[i] != val)
                nums[k++] = nums[i];

        return k;
    }

    public static void main(String[] args) {
        int[] nums = {3, 2, 2, 3};
        int val = 3;

        System.out.println((new Solution()).removeElement(nums, val));
    }
}

File: Python/_0027_Remove_Element/Solution.py
# coding=utf-8
from typing import List


# 执行用时：36 ms, 在所有 Python3 提交中击败了83.55%的用户
# 内存消耗：14.7 MB, 在所有 Python3 提交中击败了82.98%的用户
class Solution:
    def removeElement(self, nums: List[int], val: int) -> int:
        k = 0
        for i in range(len(nums)):
            if nums[i] != val:
                nums[k] = nums[i]
                k += 1
        return k


if __name__ == "__main__":
    print(Solution().removeElement([3, 2, 2, 3], 3))  # 2
